packages/shared/src/components/MainLayout.tsx

```diff
@@ -132,6 +140,24 @@ export default function MainLayout({
     );
   };
 
+  const router = useRouter();
+  const feeds = Object.values(MainFeedPage);
+  const page = router?.route?.substring(1).trim() as MainFeedPage;
+  const isPageReady = (isFeaturesLoaded && router.isReady) || isTesting;
+  const shouldShowOverlay =
+    !user &&
+    isPageReady &&
+    onboardingV2 !== OnboardingV2.Control &&
+    (!page || feeds.includes(page));
+
+  useEffect(() => {
+    if (!shouldShowOverlay) return;
+
+    router.push(`${webappUrl}/onboarding`);
+  }, [shouldShowOverlay, router]);
+
+  if (!isFeaturesLoaded || shouldShowOverlay) return null;
```

Comment:
This should never come to shouldShowOverlay right?
Because the useEffect above would already redirect.

Reply:
First render would still happen before `useEffect` is run for the first time.

Reply:
Same for updates, update to `shouldShowOverlay` would render and then `useEffect` would run.